File: BillNote_frontend/src/services/batch.ts
import request, { type IResponse } from '@/utils/request'
import toast from 'react-hot-toast'

export interface DetectedEntry {
  video_id: string
  title: string
  duration: number
  thumbnail: string
  video_url: string
}

export interface DetectUrlResponse {
  type: 'single' | 'multi'
  title?: string
  source_url?: string
  cover_url?: string
  entries: DetectedEntry[]
}

export const detectUrl = async (url: string): Promise<DetectUrlResponse> => {
  // Bilibili space list extraction may take longer due to browser automation.
  return await request.post('/detect_url', { url }, { timeout: 300000 })
}

export interface GenerateBatchRequest {
  video_urls: string[]
  platform: string
  quality: string
  model_name: string
  provider_id: string
  format: string[]
  style: string
  title?: string
  source_url?: string
  cover_url?: string
  extras?: string
  screenshot?: boolean
  link?: boolean
  video_understanding?: boolean
  video_interval?: number
  grid_size: number[]
}

export interface GenerateBatchResponse {
  batch_id: string
  task_map: Array<{ video_url: string; task_id: string }>
}

export interface BatchCourseSummary {
  batch_id: string
  title: string
  source_url: string
  cover_url: string
  total: number
  completed: number
  failed: number
  running: number
  status: string
  created_at: string
  updated_at: string
}

export interface BatchCourseItem {
  task_id: string
  video_url?: string
  title?: string
  thumbnail?: string
  duration?: number
  status?: string
  result_ready?: boolean
  note_excerpt?: string
  order?: number
  error?: string
}

export interface BatchCourseResponse extends BatchCourseSummary {
  current_task_id: string | null
  items: BatchCourseItem[]
}

export interface RawBatchTask {
  video_id?: string
  video_url?: string
  status?: string
  error?: string
  order?: number
}

export interface RawBatchEntry {
  video_url: string
  order?: number
}

export interface RawBatchStatusResponse {
  batch_id: string
  total: number
  completed: number
  failed: number
  tasks: Record<string, RawBatchTask>
  entries?: RawBatchEntry[]
  title?: string
  source_url?: string
  cover_url?: string
  created_at: string
  updated_at: string
}

export const toBatchCourseSummary = (raw: RawBatchStatusResponse): BatchCourseSummary => {
  const running = Math.max(raw.total - raw.completed - raw.failed, 0)
  const status =
    raw.total === 0 ? 'EMPTY' : running > 0 ? 'RUNNING' : raw.failed > 0 ? 'FAILED' : 'SUCCESS'

  return {
    batch_id: raw.batch_id,
    title: raw.title || '批量课程',
    source_url: raw.source_url || raw.entries?.[0]?.video_url || '',
    cover_url: raw.cover_url || '',
    total: raw.total,
    completed: raw.completed,
    failed: raw.failed,
    running,
    status,
    created_at: raw.created_at,
    updated_at: raw.updated_at,
  }
}

export interface DeleteBatchTaskResponse extends RawBatchStatusResponse {
  deleted_task_id: string
  remaining_task_ids: string[]
}

export interface DeleteBatchResponse {
  batch_id: string
  deleted_task_ids: string[]
}

export const generateBatchNote = async (data: GenerateBatchRequest): Promise<GenerateBatchResponse> => {
  return await request.post('/generate_batch_note', data, { timeout: 120000 })
}

export const getBatchStatus = async (batchId: string): Promise<RawBatchStatusResponse> => {
  return await request.get(`/batch_status/${batchId}`)
}

export const listBatchCourses = async (): Promise<BatchCourseSummary[]> => {
  return await request.get('/batch_courses')
}

export const deleteBatchTask = async (
  batchId: string,
  taskId: string
): Promise<DeleteBatchTaskResponse> => {
  try {
    const res: DeleteBatchTaskResponse = await request.post('/delete_batch_task', {
      batch_id: batchId,
      task_id: taskId,
    })
    toast.success('任务已成功删除')
    return res
  } catch (e) {
    console.error('删除批量任务失败:', e)
    throw e
  }
}

export const deleteBatch = async (batchId: string): Promise<DeleteBatchResponse> => {
  try {
    const res: DeleteBatchResponse = await request.delete(`/delete_batch/${batchId}`, {
      suppressErrorToast: true,
    })
    toast.success('合集已成功删除')
    return res
  } catch (e) {
    const response = e as IResponse | undefined
    if (response?.code === 404) {
      return { batch_id: batchId, deleted_task_ids: [] }
    }

    console.error('删除视频合集失败:', e)
    toast.error(response?.msg || '删除合集失败')
    throw e
  }
}

File: backend/app/routers/batch.py
import uuid
import json
import os
from typing import Optional

from fastapi import APIRouter, BackgroundTasks, HTTPException
from pydantic import BaseModel

from app.enmus.note_enums import DownloadQuality
from app.enmus.task_status_enums import TaskStatus
from app.services.batch_manager import BatchManager
from app.utils.response import ResponseWrapper as R
from app.utils.url_detector import UrlDetector


router = APIRouter()


def _write_status(task_id: str, status: TaskStatus, message: str = '') -> None:
    output_dir = os.getenv('NOTE_OUTPUT_DIR', 'note_results')
    os.makedirs(output_dir, exist_ok=True)
    path = os.path.join(output_dir, f'{task_id}.status.json')
    with open(path, 'w', encoding='utf-8') as f:
        json.dump({'status': status.value, 'message': message}, f, ensure_ascii=False, indent=2)


class DetectUrlRequest(BaseModel):
    url: str


class BatchGenerateRequest(BaseModel):
    video_urls: list[str]
    platform: str
    quality: DownloadQuality
    model_name: str
    provider_id: str
    title: Optional[str] = None
    source_url: Optional[str] = None
    cover_url: Optional[str] = None
    screenshot: Optional[bool] = False
    link: Optional[bool] = False
    format: Optional[list] = []
    style: Optional[str] = None
    extras: Optional[str] = None
    video_understanding: Optional[bool] = False
    video_interval: Optional[int] = 0
    grid_size: Optional[list] = []


class DeleteBatchTaskRequest(BaseModel):
    batch_id: str
    task_id: str


@router.post('/detect_url')
def detect_url(data: DetectUrlRequest):
    try:
        res = UrlDetector.detect(data.url)
        return R.success(data=res)
    except Exception as e:
        return R.error(msg=str(e))


@router.post('/generate_batch_note')
def generate_batch_note(data: BatchGenerateRequest, background_tasks: BackgroundTasks):
    if not data.video_urls:
        raise HTTPException(status_code=400, detail='video_urls is empty')

    mgr = BatchManager()
    batch_id = mgr.create_batch(
        data.video_urls,
        title=data.title or '',
        source_url=data.source_url or data.video_urls[0],
        cover_url=data.cover_url or '',
    )
    task_map: list[dict] = []

    # Schedule each task independently in the shared executor.
    from app.services.task_serial_executor import task_serial_executor
    from app.routers import note as note_router

    for video_url in data.video_urls:
        task_id = str(uuid.uuid4())
        _write_status(task_id, TaskStatus.PENDING)
        mgr.register_task(batch_id, task_id, video_url)
        task_map.append({'video_url': video_url, 'task_id': task_id})

        task_serial_executor.submit(
            note_router._run_note_task_impl,
            task_id,
            video_url,
            data.platform,
            data.quality,
            data.link,
            data.screenshot,
            data.model_name,
            data.provider_id,
            data.format,
            data.style,
            data.extras,
            data.video_understanding,
            data.video_interval,
            data.grid_size,
        )

    return R.success(data={'batch_id': batch_id, 'task_map': task_map})


@router.get('/batch_courses')
def batch_courses():
    mgr = BatchManager()
    return R.success(data=mgr.list_batches())


@router.post('/delete_batch_task')
def delete_batch_task(data: DeleteBatchTaskRequest):
    try:
        mgr = BatchManager()
        result = mgr.delete_task_from_batch(data.batch_id, data.task_id)
        ordered_task_ids = sorted(
            result.get('tasks', {}).keys(),
            key=lambda task_id: (result['tasks'][task_id].get('order', 0), task_id),
        )
        return R.success(
            data={
                **result,
                'deleted_task_id': data.task_id,
                'remaining_task_ids': ordered_task_ids,
            },
            msg='删除成功',
        )
    except FileNotFoundError as e:
        missing_id = e.args[0] if e.args else str(e)
        message = 'batch not found' if missing_id == data.batch_id else 'task not found'
        return R.error(msg=message, code=404)
    except ValueError as e:
        return R.error(msg=str(e), code=400)


@router.delete('/delete_batch/{batch_id}')
def delete_batch(batch_id: str):
    try:
        mgr = BatchManager()
        return R.success(data=mgr.delete_batch(batch_id), msg='删除成功')
    except FileNotFoundError:
        return R.error(msg='batch not found', code=404)
    except ValueError as e:
        return R.error(msg=str(e), code=400)


@router.get('/batch_status/{batch_id}')
def batch_status(batch_id: str):
    try:
        mgr = BatchManager()
        return R.success(data=mgr.refresh_from_task_status(batch_id))
    except FileNotFoundError:
        return R.error(msg='batch not found', code=404)
    except Exception as e:
        return R.error(msg=str(e))
